src/util.rs

```diff
@@ -0,0 +1,15 @@
+/// Utility functions.
+
+/// Returns true if a slice is a permutation.
+///
+/// That is, all the elements in `0..len` occur exactly once in the slice.
+pub fn is_permutation(v: &[usize]) -> bool {
+    let n = v.len();
+    let mut seen = (0..n).map(|_| false).collect::<Vec<_>>();
+    for &e in v {
+        if (0..n).contains(&e) {
```

Comment:
Since this is an unsigned value, I think we can just check

```suggestion
        if e < n {
```

Reply:
Though see my other suggestion about early return.